### PR TITLE
added caps:a

### PR DESCRIPTION
**Pull request by HookedBehemoth**
I'm almost certain some of these are deprecated on newer versions.

### CODE REVIEW

File: nx/include/switch/services/capsa.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
+/**
+ * @file capsu.h
```
Comment (yellows8) [inline nx/include/switch/services/capsa.h]: Wrong filename.

File: nx/include/switch/services/capsa.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    AlbumObjectLocation_SdCard      = 1,
+} AlbumObjectLocation;
+
+/// Initialize caps:a.
+Result capsaInitialize(void);
+
+/// Exit caps:a.
+void capsaExit(void);
+
+/// Gets the Service for caps:u.
```
Comment (yellows8) [inline nx/include/switch/services/capsa.h]: Some of these have the wrong prefix.

File: nx/include/switch/services/capsa.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * @brief Album Accessor (caps:a) service IPC wrapper.
+ * @author Behemoth
+ * @copyright libnx Authors
+ */
+#pragma once
+#include "../types.h"
+#include "../sf/service.h"
+#include "../services/caps.h"
+
+typedef enum {
```
Comment (yellows8) [inline nx/include/switch/services/capsa.h]: Should be in caps.h (where's the name from)?

Comment (HookedBehemoth) [reply nx/include/switch/services/capsa.h]: just a guess

**Comment by HookedBehemoth** [comment]
I can't edit the capsrv wiki page. :/
`*** Forbidden. Message seems to be spam. ***`

**Comment by mha42** [comment]
> I can't edit the capsrv wiki page. :/
> `*** Forbidden. Message seems to be spam. ***`

Your account, mail and current IP during post has been whitelisted. Also posted the contents to cleantalk to train the service.

File: nx/include/switch/services/caps.h [hunk 4]
```diff
@@ -78,7 +78,7 @@ typedef struct {
 typedef struct {
     u64 program_id;                              ///< ProgramId.
     CapsAlbumFileDateTime datetime;              ///< \ref CapsAlbumFileDateTime
-    u8 unk_x10;                                  ///< Unknown.
+    u8 location;                                 ///< AlbumLocation
     u8 unk_x11;                                  ///< Unknown.
     u8 pad[6];                                   ///< Padding?
 } CapsAlbumEntryId;
```
Comment (SciresM) [inline nx/include/switch/services/caps.h]: Actual name for this struct should be `CapsAlbumFileId`.

File: nx/include/switch/services/caps.h [hunk 5]
```diff
@@ -78,7 +78,7 @@ typedef struct {
 typedef struct {
     u64 program_id;                              ///< ProgramId.
     CapsAlbumFileDateTime datetime;              ///< \ref CapsAlbumFileDateTime
-    u8 unk_x10;                                  ///< Unknown.
+    u8 location;                                 ///< AlbumLocation
     u8 unk_x11;                                  ///< Unknown.
```
Comment (SciresM) [inline nx/include/switch/services/caps.h]: ///< FileContentsType

File: nx/include/switch/services/caps.h [hunk 6]
```diff
@@ -78,7 +78,7 @@ typedef struct {
 typedef struct {
     u64 program_id;                              ///< ProgramId.
```
Comment (SciresM) [inline nx/include/switch/services/caps.h]: maybe rename this to application_id

File: nx/include/switch/services/capsa.h [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    AlbumObjectLocation_BuiltInUser = 0,
+    AlbumObjectLocation_SdCard      = 1,
+} AlbumObjectLocation;
```
Comment (SciresM) [inline nx/include/switch/services/capsa.h]: this is `nn::capsrv::AlbumStorage`, so:

```c
typedef enum {
    CapsAlbumStorage_Nand = 0,
    CapsAlbumStorage_Sd   = 1,
} CapsAlbumStorage;

File: nx/include/switch/services/capsa.h [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+/// Exit caps:a.
+void capsaExit(void);
+
+/// Gets the Service for caps:u.
+Service* capsuGetServiceSession(void);
+
+Result capsaGetAlbumFileCount(AlbumObjectLocation location, u64* count);
+Result capsaGetAlbumFileList(AlbumObjectLocation location, u64* count, CapsApplicationAlbumEntry* buffer, u64 size);
+Result capsaLoadAlbumFile(CapsAlbumEntryId entry_id, u8 unk[8], void* jpeg_buffer, u64 jpeg_buffer_size);
```
Comment (SciresM) [inline nx/include/switch/services/capsa.h]:
```c
Result capsaLoadAlbumFile(CapsAlbumEntryId entry_id, u64 *out_size, void* jpeg_buffer, u64 jpeg_buffer_size);
```

File: nx/include/switch/services/capsa.h [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+Result capsaDeleteAlbumFile(CapsAlbumEntryId entry_id);
+Result capsaStorageCopyAlbumFile(u8 unk[0x20]);
```
Comment (SciresM) [inline nx/include/switch/services/capsa.h]:
```c
Result capsaStorageCopyAlbumFile(CapsAlbumFileId file_id, u8 CapsAlbumStorageType dst_storage);
```

File: nx/include/switch/services/capsa.h [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+Result capsaIsAlbumMounted(AlbumObjectLocation location, bool* is_mounted);
+Result capsaGetAlbumUsage(AlbumObjectLocation location, u8 unk_out[0x30]);
```
Comment (SciresM) [inline nx/include/switch/services/capsa.h]:
```c
typedef enum {
    CapsAlbumFileContents_ScreenShot      = 0,
    CapsAlbumFileContents_Movie           = 1,
    CapsAlbumFileContents_ExtraScreenShot = 2,
    CapsAlbumFileContents_ExtraMovie      = 3,
} CapsAlbumFileContents;

typedef enum {
    CapsAlbumContentsUsageFlag_HasGreaterUsage   = (1 << 0), ///< Indicates that there are additional files not captured by the count/size fields of CapsAlbumContentsUsage
    CapsAlbumContentsUsageFlag_IsUnknownContents = (1 << 1), ///< Indicates that the file is not a known content type 
} CapsAlbumContentsUsageFlag;

typedef struct {
    s64 count;
    s64 size;
    u32 flags;
    u8 file_contents;
} CapsAlbumContentsUsage;

typedef struct {
    CapsAlbumContentsUsage usages[2];
} CapsAlbumUsage2;

typedef struct {
    CapsAlbumContentsUsage usages[3];
} CapsAlbumUsage3;

typedef struct {
    CapsAlbumContentsUsage usages[16];
} CapsAlbumUsage16;

Result capsaGetAlbumUsage(CapsAlbumStorage storage, CapsAlbumUsage2 *out);
```

File: nx/include/switch/services/capsa.h [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+Result capsaGetAlbumFileSize(CapsAlbumEntryId entry_id, u64* size);
+Result capsLoadAlbumFileThumbnail(CapsAlbumEntryId entry_id, u8 unk[8], void* jpeg_buffer, u64 jpeg_buffer_size);
```
Comment (SciresM) [inline nx/include/switch/services/capsa.h]:
```c
Result capsLoadAlbumFileThumbnail(CapsAlbumEntryId entry_id, u64 *out_size, void* jpeg_buffer, u64 jpeg_buffer_size);
```

**Comment by HookedBehemoth** [comment]
Thanks a lot! I started reversing capsrv after I did the commit but only found AlbumFileId on my own. ^^

**Comment by HookedBehemoth** [comment]
LoadAlbumScreenShotThumbnailImage, -Ex, -Ex1 all fill the unsecure buffer with raw pixel values and leave the jpeg_buffer clean. They still return 0.

Internally it calls a function pointer at 710014e668. I'd assume that there should be something from jpegdec but I'm not sure...

**Comment by yellows8** [comment]
"jpeg_buffer ..." Probably (?) the same as the workbuf used by capsu.

**Comment by SciresM** [comment]
Yeah, that arg is used ass a workbuffer.

File: nx/source/services/capsa.c [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+Result capsaGetAlbumUsage3(CapsAlbumStorage storage, CapsAlbumUsage3 *out) {
+    if (hosversionBefore(4,0,0))
+        return MAKERESULT(Module_Libnx, LibnxError_IncompatSysVer);
+    return serviceDispatchInOut(&g_capsaSrv, 15, storage, *out);
+}
+
+Result capsaGetAlbumMountResult(CapsAlbumStorage storage) {
+    if (hosversionBefore(4,0,0))
+        return MAKERESULT(Module_Libnx, LibnxError_IncompatSysVer);
+    return serviceDispatchIn(&g_capsaSrv, 17, storage);
```
Comment (yellows8) [inline nx/source/services/capsa.c]: cmdid typo

File: nx/source/services/capsa.c [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+void _capsaCleanup(void) {
+    serviceClose(&g_capsaSrv);
+}
+
+Service* capsaGetServiceSession(void) {
+    return &g_capsaSrv;
+}
+
+Result capsaGetAlbumFileCount(CapsAlbumStorage storage, u64* count) {
+    return serviceDispatchInOut(&g_capsaSrv, 0, storage, *count);
```
Comment (yellows8) [inline nx/source/services/capsa.c]: In IPC CapsAlbumStorage is an u8, so store it in a tmp u8 field (likewise for everything else using it).

File: nx/source/services/capsa.c [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+Result capsaGetAlbumFileList(CapsAlbumStorage storage, u64* count, CapsApplicationAlbumEntry* buffer, u64 buffer_size) {
+    return serviceDispatchInOut(&g_capsaSrv, 1, storage, *count,
+        .buffer_attrs = { SfBufferAttr_HipcMapAlias | SfBufferAttr_Out },
+        .buffers = { { buffer, buffer_size } },
+    );
+}
+
+Result capsaLoadAlbumFile(CapsAlbumFileId file_id, u64 *out_size, void* jpeg_buffer, u64 jpeg_buffer_size) {
```
Comment (yellows8) [inline nx/source/services/capsa.c]: workbuf(_size), same elsewhere.

File: nx/source/services/capsa.c [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+
+Result capsaGetRequiredStorageSpaceSizeToCopyAll(CapsAlbumStorage dst_storage, CapsAlbumStorage src_storage, u64* out) {
+    struct {
+        CapsAlbumStorage dest;
+        CapsAlbumStorage src;
+    } in = { dst_storage, src_storage };
+    return serviceDispatchInOut(&g_capsaSrv, 501, in, *out);
+}
+
+Result capsaLoadAlbumScreenShotThumbnailImage(u64* width, u64* height, CapsAlbumFileId file_id, void* jpeg_buffer, u64 jpeg_buffer_size, void* buffer, u64 buffer_size) {
```
Comment (yellows8) [inline nx/source/services/capsa.c]: Function order should match cmdid order (same with the .h).

File: nx/source/services/capsa.c [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+        .buffers = { { jpeg_buffer, jpeg_buffer_size } },
+    );
+}
+
+Result capsaDeleteAlbumFile(CapsAlbumFileId file_id) {
+    return serviceDispatchIn(&g_capsaSrv, 3, file_id);
+}
+
+Result capsaStorageCopyAlbumFile(CapsAlbumFileId file_id, CapsAlbumStorage dst_storage) {
+    return serviceDispatchIn(&g_capsaSrv, 4, dst_storage);
```
Comment (yellows8) [inline nx/source/services/capsa.c]: Guess an in-struct is missing?

File: nx/source/services/capsa.c [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+
+Result capsaLoadAlbumFileThumbnail(CapsAlbumFileId file_id, u64 *out_size, void* jpeg_buffer, u64 jpeg_buffer_size) {
+    return serviceDispatchInOut(&g_capsaSrv, 8, file_id, *out_size,
+        .buffer_attrs = { SfBufferAttr_HipcMapAlias | SfBufferAttr_Out },
+        .buffers = { { jpeg_buffer, jpeg_buffer_size } },
+    );
+}
+
+Result capsaGetAlbumUsage3(CapsAlbumStorage storage, CapsAlbumUsage3 *out) {
+    if (hosversionBefore(4,0,0))
```
Comment (yellows8) [inline nx/source/services/capsa.c]: Doc all required system-versions in the .h.

File: nx/include/switch/services/capsa.h [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+
+/// Initialize caps:a.
+Result capsaInitialize(void);
+
+/// Exit caps:a.
+void capsaExit(void);
+
+/// Gets the Service for caps:a.
+Service* capsaGetServiceSession(void);
+// 0-8
```
Comment (yellows8) [inline nx/include/switch/services/capsa.h]: No need for cmdid comments in the .h.

File: nx/source/services/capsa.c [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+    return rc;
+}
+
+Result capsaLoadAlbumScreenShotThumbnailImageEx(u64* width, u64* height, CapsAlbumFileId file_id, void* jpeg_buffer, u64 jpeg_buffer_size, void* buffer, u64 buffer_size) {
+    if (hosversionBefore(3,0,0))
+        return MAKERESULT(Module_Libnx, LibnxError_IncompatSysVer);
+    struct {
+        CapsAlbumFileId file_id;
+        CapsScreenShotDecodeOption opts;
+    } in = { file_id, {{0}} };
```
Comment (yellows8) [inline nx/source/services/capsa.c]: opts should be user-specified, same elsewhere.

File: nx/source/services/capsa.c (hunk 12 above)
Comment (yellows8) [inline nx/source/services/capsa.c]: There's a bunch of dup funcs for this, use something like: `static Result _capsaCmdInU8NoOut(Service* srv, u8 inval, u32 cmd_id)`

**Comment by HookedBehemoth** [comment]
Shouldn't CapsAlbumContentsUsage have padding to make sure it's always 0x18 big?

**Comment by yellows8** [comment]
sure

**Comment by HookedBehemoth** [comment]
Seams to me like CapsApplicationAlbumEntry is actually CapsAlbumEntry.
Where does the diff from the versions come from?

**Comment by yellows8** [comment]
Those are different, see elsewhere.

File: nx/include/switch/services/caps.h [hunk 20]
```diff
@@ -22,6 +22,11 @@ typedef enum {
     AlbumReportOption_Enable = 1,              ///< Display the screenshot-taken Overlay notification.
 } AlbumReportOption;
 
+typedef enum {
+    CapsAlbumStorage_Nand = 0,        ///< BuiltInUser
```
Comment (yellows8) [inline nx/include/switch/services/caps.h]: Nand/Sd for the comments.

File: nx/include/switch/services/caps.h [hunk 21]
```diff
@@ -76,17 +81,18 @@ typedef struct {
...
 /// AlbumEntryId
 typedef struct {
-    u64 program_id;                              ///< ProgramId.
+    u64 application_id;                          ///< ApplicationId
     CapsAlbumFileDateTime datetime;              ///< \ref CapsAlbumFileDateTime
-    u8 unk_x10;                                  ///< Unknown.
-    u8 unk_x11;                                  ///< Unknown.
-    u8 pad[6];                                   ///< Padding?
-} CapsAlbumEntryId;
+    u8 storage;                                  ///< AlbumStorage
```
Comment (yellows8) [inline nx/include/switch/services/caps.h]: \ref CapsAlbumStorage

File: nx/include/switch/services/caps.h [hunk 22]
```diff
@@ -76,17 +81,18 @@ typedef struct {
...
+    u8 content;                                  ///< FileContentsType
```
Comment (yellows8) [inline nx/include/switch/services/caps.h]: \ref CapsAlbumFileContents

File: nx/include/switch/services/caps.h [hunk 23]
```diff
@@ -76,17 +81,18 @@ typedef struct {
...
+    u32 pad_x12;                                 ///< Set to 0 by official software
+    u16 pad_x16;                                 ///< Set to 0 by official software
+} PACKED CapsAlbumFileId;
 
 /// AlbumEntry
 typedef struct {
-    u8 unk_x0[0x8];                              ///< Unknown.
-    CapsAlbumEntryId id;                         ///< \ref CapsAlbumEntryId
+    u64 size;                                    ///< Unknown.
```
Comment (yellows8) [inline nx/include/switch/services/caps.h]: Update comment?

File: nx/include/switch/services/caps.h [hunk 24]
```diff
@@ -120,6 +126,38 @@ typedef struct {
     u32 size;                                    ///< UserData size.
 } CapsApplicationData;
 
+typedef enum {
```
Comment (yellows8) [inline nx/include/switch/services/caps.h]: Before these typedefs add: `/// {name without the Caps prefix}` (same above)

File: nx/include/switch/services/capsa.h [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+
+/// Initialize caps:a.
+Result capsaInitialize(void);
+
+/// Exit caps:a.
+void capsaExit(void);
+
+/// Gets the Service for caps:a.
+Service* capsaGetServiceSession(void);
+Result capsaGetAlbumFileCount(CapsAlbumStorage storage, u64* count);
```
Comment (yellows8) [inline nx/include/switch/services/capsa.h]: Add newline after capsaGetServiceSession.

File: nx/source/services/capsa.c [hunk 26]
```diff
@@ -0,0 +1,190 @@
...
+
+Result capsaGetAlbumUsage16(CapsAlbumStorage storage, CapsAlbumUsage16 *out) {
+    if (hosversionBefore(4,0,0))
+        return MAKERESULT(Module_Libnx, LibnxError_IncompatSysVer);
+    u8 inval = storage;
+    return serviceDispatchInOut(&g_capsaSrv, 17, inval, *out);
+}
+
+Result capsaGetAutoSavingStorage(CapsAlbumStorage* storage) {
+    return serviceDispatchOut(&g_capsaSrv, 401, *storage);
```
Comment (yellows8) [inline nx/source/services/capsa.c]: u8 tmpval for output.

File: nx/include/switch/services/capsa.h [hunk 27]
```diff
@@ -0,0 +1,47 @@
...
+Result capsaStorageCopyAlbumFile(CapsAlbumFileId file_id, CapsAlbumStorage dst_storage);
+Result capsaIsAlbumMounted(CapsAlbumStorage storage, bool* is_mounted);
+Result capsaGetAlbumUsage(CapsAlbumStorage storage, CapsAlbumUsage2 *out);
+Result capsaGetAlbumFileSize(CapsAlbumFileId file_id, u64* size);
+Result capsaLoadAlbumFileThumbnail(CapsAlbumFileId file_id, u64 *out_size, void* workbuf, u64 workbuf_size);
+/// Only available on [2.0.0+].
+Result capsaLoadAlbumScreenShotImage(u64* width, u64* height, CapsAlbumFileId file_id, void* workbuf, u64 workbuf_size, void* rawbuf, u64 rawbuf_size);
+Result capsaLoadAlbumScreenShotThumbnailImage(u64* width, u64* height, CapsAlbumFileId file_id, void* workbuf, u64 workbuf_size, void* rawbuf, u64 rawbuf_size);
+/// Only available on [3.0.0+].
+Result capsaLoadAlbumScreenShotImageEx(u64* width, u64* height, CapsAlbumFileId file_id, CapsScreenShotDecodeOption opts, void* workbuf, u64 workbuf_size, void* rawbuf, u64 rawbuf_size);
```
Comment (yellows8) [inline nx/include/switch/services/capsa.h]: const CapsScreenShotDecodeOption *opts, same below.

File: nx/include/switch/services/capsa.h [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+Result capsaInitialize(void);
+
+/// Exit caps:a.
+void capsaExit(void);
+
+/// Gets the Service for caps:a.
+Service* capsaGetServiceSession(void);
+Result capsaGetAlbumFileCount(CapsAlbumStorage storage, u64* count);
+Result capsaGetAlbumFileList(CapsAlbumStorage storage, u64* count, CapsAlbumEntry* buffer, u64 size);
+Result capsaLoadAlbumFile(CapsAlbumFileId file_id, u64 *out_size, void* workbuf, u64 workbuf_size);
```
Comment (yellows8) [inline nx/include/switch/services/capsa.h]: const CapsAlbumFileId *file_id, same below

**Pull request merged**